src/stcal/resample/resample.py

```diff
@@ -0,0 +1,1506 @@
+    if bbox is None:
+        xmin = ymin = 0
+        xmax = data_shape[1] - 1
+        ymax = data_shape[0] - 1
+    else:
+        ((x1, x2), (y1, y2)) = bbox
+        xmin = max(0, int(x1 + 0.5))
+        ymin = max(0, int(y1 + 0.5))
+        xmax = min(data_shape[1] - 1, int(x2 + 0.5))
+        ymax = min(data_shape[0] - 1, int(y2 + 0.5))
+
+    return xmin, xmax, ymin, ymax
+
+
+class ResampleModelIO(abc.ABC):
+    @abc.abstractmethod
+    def open_model(self, file_name):
+        ...
+
+    @abc.abstractmethod
+    def get_model_attr_value(self, model, attribute_name):
+        ...
+
+    @abc.abstractmethod
+    def set_model_attr_value(self, model, attribute_name, value):
+        ...
+
+    @abc.abstractmethod
+    def get_model_meta(self, model, attributes):
+        ...
+
+    @abc.abstractmethod
+    def set_model_meta(self, model, attributes):
+        ...
+
+    @abc.abstractmethod
+    def get_model_array(self, model, array_name):
+        ...
+
+    @abc.abstractmethod
+    def set_model_array(self, model, array_name, data):
+        ...
+
+    @abc.abstractmethod
+    def close_model(self, model):
+        ...
+
+    @abc.abstractmethod
+    def save_model(self, model):
+        ...
+
+    @abc.abstractmethod
+    def write_model(self, model, file_name, **kwargs):
+        ...
+
+    @abc.abstractmethod
+    def new_model(self, image_shape=None, file_name=None, copy_meta_from=None):
+        """ Return a new model for the resampled output """
+        ...
+
+
+class OutputTooLargeError(RuntimeError):
+    """Raised when the output is too large for in-memory instantiation"""
+
+
+def output_wcs_from_input_wcs(input_wcs_list, pixel_scale_ratio=1.0,
+                              pixel_scale=None, output_shape=None,
+                              crpix=None, crval=None, rotation=None):
+    # TODO: should be replaced with a version that lives in stcal and
+    # uses s_region
+    w = deepcopy(input_wcs_list[0])  # this is bad
+    return {
+        'output_wcs': w,
+        'pscale': np.rad2deg(np.sqrt(_compute_image_pixel_area(w))),
+        'pscale_ratio': 1.0,
+        'crpix': None
+    }
+
+
+class ResampleBase(abc.ABC):
+    """
+    This is the controlling routine for the resampling process.
+
+    Notes
+    -----
+    This routine performs the following operations::
+
+      1. Extracts parameter settings from input model, such as pixfrac,
+         weight type, exposure time (if relevant), and kernel, and merges
+         them with any user-provided values.
+      2. Creates output WCS based on input images and define mapping function
+         between all input arrays and the output array.
+      3. Updates output data model with output arrays from drizzle, including
+         a record of metadata from all input models.
+    """
+    resample_suffix = 'i2d'
+    resample_file_ext = '.fits'
+    n_arrays_per_output = 2  # #flt-point arrays in the output (data, weight, var, err, etc.)
```

Comment:
why is this hard-coded?  Will it be different for co-add and single resamples?  Don't we have multiple types of variance that need to be in memory simultaneously in the current version of resample, i.e., doesn't this underestimate the memory usage when it's used in `check_memory_requirements`?